Perfect FreeType-based outlines for DynamicFonts

endragor opened the pull request:
- Implement outlines based on FreeType Stroker API. This allows
  artifact-free results, similar to what you will see in Web or any text
  editing tools. Outline parameters are a part of DynamicFont rather
  than Label, because outlines have to be baked into the font's atlas
  along with the characters.

- Add optional no_outline parameter to Font::draw_char so that shadows
  can be rendered as expected - without including the outline.

- Add start_draw_batch/end_draw_batch methods which are supposed to be
  called before/after a series of draw calls. This is needed because all
  outlines in a text have to be rendered before the characters. This
  also allows potential optimizations for text rendering.

- Number of draw-calls is reduced from 5 to 2 per outlined character.

- Overall cleanup of DynamicFont code, extracted duplicated code pieces
  into separate methods.

- The change is backward-compatible - Labels still have outline
  properties that work exactly as they worked before.

Closes #16279.

endragor committed:
Perfect FreeType-based outlines for DynamicFonts

- Implement outlines based on FreeType Stroker API. This allows
  artifact-free results, similar to what you will see in Web or any text
  editing tools. Outline is a part of DynamicFont rather than Label,
  because outlines have to be baked into the font's atlas. Font has a
  default outline_color and a Label can specify font_outline_modulator
  that will be multiplied with the Font's color to get the final result.

- draw_char now has to be called twice to fully render a text - first
  with p_outline == true for each character and then with
  p_outline == false for each character.

- Number of draw-calls is reduced from 5 to 2 per outlined character.

- Overall cleanup of DynamicFont code, extracted duplicated code pieces
  into separate methods.

- The change is backward-compatible - Labels still have outline
  properties that work exactly as they worked before.

Closes #16279.

akien-mga commented on scene/resources/dynamic_font.cpp:
```diff
@@ -747,6 +801,29 @@ int DynamicFont::get_size() const {
+void DynamicFont::set_outline_size(int p_size) {
+	if (cache_id.outline_size == p_size)
+		return;
+	ERR_FAIL_COND(p_size < 0 || p_size > UINT16_MAX);
```
Doesn't build on MSVC 2015, likely a missing include: https://ci.appveyor.com/project/akien-mga/godot/build/1.0.6831

akien-mga commented on scene/resources/dynamic_font.cpp:
```diff
@@ -33,9 +33,20 @@
 #include "os/os.h"
 
-bool DynamicFontData::CacheID::operator<(CacheID right) const {
+#include <freetype/ftstroke.h>
```
This fails building on Travis (Ubuntu 14.04) against the system freetype2, as they put freetype headers in `/usr/include/freetype2/ftstroke.h` instead of the standard `/usr/include/freetype2/freetype/ftstroke.h`.

I believe that's why freetype devs expect you to use their include macros defined in `ftheader.h`: https://www.freetype.org/freetype2/docs/reference/ft2-header_file_macros.html

IIUC, to include freetype stuff you can only refer to `ft2build.h` as entry point (which includes `ftheader.h`), and then you can use `#include FT_STROKER_H`. Here `ft2build.h` is already included in `dynamic_font.h`, so you only need the latter: https://github.com/endragor/godot/blob/3eb7eeed7968bccc79a933626418d8e9602a6023/scene/resources/dynamic_font.h#L41

endragor committed:
Always emit dynamic font change in update_oversampling

Fixes #15787.

The issue occurred when two (or more) separate DynamicFont instances
used the same DynamicFontAtSize instance due to having equal
properties. The first instance updated its data_at_size and emitted
"changed" signal, but the second did not because it considered the
data_at_size to be up to date, even though it has just been updated.

endragor commented:
I've added a fix to #15787 as a separate commit to this branch since it requires a change to dynamic_font.cpp which was heavily refactored here.

reduz commented:
I feel hashing the font color is way too much overkill and allows itself to be heavily misused.
I would only either:
* Just allow black outlines
* Find a way to draw the font outline separately (as in, only outline) so they can be composited.

reduz commented:
I am pretty sure it should be possible to draw outlines only, right? if we can separate outline from font, we can composite them on demand and for different colors if needed.

endragor commented:
Could you please clarify how hasing is an overkill and could be misused, especially when it's local to the font type? I don't get it.

reduz commented:
The problem is not hashing itself, but having a font version for every outline color. What I wonder is if the outline can be composited in two draw calls to avoid this (both have a caret for char and for outline and draw them together).

endragor commented:
Updated the PR based on discussions with @reduz.

- There are now separate DynamicFontAtSize instances created for base font and its outlines, which means there will be no additional memory usage if the same font has multiple different outline sizes. Having different outline colors also does not affect the amount of memory used.

- Batching is now done externally to the Font object rather than internally, which means it's a responsibility of clients to call `draw_char` twice - first to render an outline for the whole text and then to render the characters themselves. `FontDrawer` was added to help with this in engine code, but scripting clients still have to take care of that. Should not be a problem though since using Font::draw_char from a script directly is a rare use case.

- `font_outline_modulator` property was added to Label class - it's used as a multiplier for the font's own outline color. Is the name OK?

endragor commented:
@reduz I've updated it based on what we agreed on - see my previous comment. I've now also resolved the conflicts which occurred after that.

Calinou commented:
> - font_outline_modulator property was added to Label class - it's used as a multiplier for the font's own outline color. Is the name OK?

`font_outline_modulate` would sound more consistent with the `modulate` and `self_modulate` properties we currently have in Node.

endragor commented:
True, modulate is used as a color property name in few places, even though it's a verb and not a noun. I've updated the property name to `font_outline_modulate` for consistency.

akien-mga commented:
Thanks!

Nobelix commented:
Could this be cherry picked for 3.0.3? As it fixes the issue with oversampling #15787 and that one is very annoying when you want to have multiple resolutions

mhilbrunner commented:
May be too late for 3.0.3, but maybe @hpvb knows better. :P I'll add the cherrypick label regardless, maybe there will be a 3.0.4 before 3.1.

AlessioCorridori commented:
Hi, I'm using
Godot Version: 3.06.stable.official.8314054
Windows Version: Windows 10 64-bit on Surface 4 Pro (with HDPI monitor)

Today, I've experimented the issue #15787 (DynamicFont occasionally broken with oversampling enabled). 
Not sure if this is the correct place to report this, but I noticed that it happens when:
- Set a custom font and save as .tres file
- Set another custom font with the exact same settings (font file and size) and save it in another .tres file
- create a scene with two labels
- assign the first .tres custom font preset to the first label
- assign the second .tres custom font present to the second label (basically, you have two labels in the same scene that read the custom font preset from two different files that has the same settings)
- with "use oversampling" option activated, when I try to scale the window the #15787 error appears, and the debugger/errors tab in the editor, list hundreds of this error:

0:00:04:0453 - Condition ' !id_map.has(p_rid.get_data()) ' is true. returned: __null
Type:Error
Description: 
Time: 0:00:04:0453
C Error: Condition ' !id_map.has(p_rid.get_data()) ' is true. returned: __null
C Source: core/rid.h:165
C Function: getornull

The error disappears:
1. Disabling the "use oversampling" option, but also
2. Just setting different font size in one of the .tres custom font present (if the two presets have the same font but different size everything seems works fine)

Hope this can help (and sorry for my bad English)

akien-mga commented:
Now that Godot 3.1 has been released, we don't plan to cherry-pick new features and enhancements to the 3.0 branch, unless there is very strong support in favor of it. Removing cherrypick label for 3.0.
